usr/src/boot/efi/libefi/efi_console.c

```diff
@@ -457,8 +457,13 @@ efi_framebuffer_setup(void)
 	}
 	shadow_sz = EFI_SIZE_TO_PAGES(efifb.fb_width * efifb.fb_height *
 	    sizeof (*shadow_fb));
+#if defined(__aarch64__)
```

Comment:
This needs a comment. on x86, we allocate below max address to avoid memory map fragmentation, because we do need space for loading kernel and modules and so far, the usable space has been limited to 4GB line (because historically there have been machines with memory mapped only up to 4GB). Today, there are cases where it would be helpful to be able to use more memory - mostly because of memory fragmentation and the need to use huge "miniroot" boot archives. But, while we only can use dboot to start the kernel, we are still bound to 4GB limit because dboot is running in 32-bit protected mode.

Reply:
Yeah, this is all processor-specific in the UEFI spec. The x64 portion (UEFI 2.11, §2.3.4) implies that all memory presented in the UEFI memory map must be identity mapped. I guess this is what happens when there's no certification process in place.

aarch64 (UEFI2.11, §2.3.6) says effectively the same thing, but with clearer wording - and there's a compliance suite that a vendor must pass, which helps to keep them honest. The only vagueness is when PA size is greater than VA size, but that's extremely unlikely.